dnprogs/contrib/ph3-der-loewe/man1/dnetcat.1
.TH "dnetcat" "1" "September 2008" "dntools" "User Commands"

.SH NAME

dnetcat \- opens a DECnet connection

.SH SYNOPSIS

dnetcat {\-h|\-\-help}

dnetcat [\-v] [\-z] node[::|[::] ]object

dnetcat [\-v] [\-f] \-l object

.SH "DESCRIPTION"
\fBdnetcat\fR is a \fBnetcat\fR(1) like tool that opens a DECnet connection and
copies all data between the standard input/output to/from the socket.

.SH OPTIONS
.SS "\-v"
Be verbose. Prints useful (debug) messages.
.SS "\-z"
Zero IO mode. In this mode a connection is dropped directly after it is opened.
This can be used to scan if the remote side listens on the given address and object.
.SS "\-l"
Listen mode. In this mode \fBdnetcat\fR accepts connections on the given object.
.SS "\-f"
Fork on new connection. This let netcat accept multiple connections in listen mode.
This is ignored if \fB\-l\fR is not given.
.SS "\-\-help, \-h"
Prints a small help.

.SH EXAMPLES
This program can be used as 'ProxyCommand' for \fBssh\fR(1).
You may use it via command line like this:
 ssh \-o 'ProxyCommand dnetcat %h ssh' [...] mynode

or you may use it in your \fB~/.ssh/config\fR like this:
 Host *::
 ProxyCommand dnetcat %h ssh

This will enable you to connect to any DECnet node by adding "::" at
the end of the node name like this:
 ssh mynode::

For more information on \fBssh\fR(1) see it's manpage.

.SH SEE ALSO
\fBssh\fR(1),
\fBnetcat\fB(1).

.SH AUTHORS
This program is written by Philipp "ph3-der-loewe" Schafft <lion@lion.leolix.org>.


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/contrib/ph3-der-loewe/dnetcat.c
//dnetcat.c:

/*
    copyright 2008 Philipp 'ph3-der-loewe' Schafft <lion@lion.leolix.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    version 3.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>

#include <signal.h>

// to support old standards:
#include <sys/time.h>
#include <sys/types.h>

// currrent standard:
#include <sys/select.h>

#include <sys/socket.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>

char * progname = NULL;

void usage (void) {
 fprintf(stderr, "Usage:\n");
 fprintf(stderr, "%s [OPTIONS] node[::] object\n", progname);
 fprintf(stderr, "or\n");
 fprintf(stderr, "%s [OPTIONS] node::object\n", progname);
 fprintf(stderr, "or\n");
 fprintf(stderr, "%s [OPTIONS] -l object\n", progname);
 fprintf(stderr, "\n");
 fprintf(stderr, "Options:\n");
 fprintf(stderr, " -h   --help        this help\n"
                 " -v                 be verbose\n"
                 " -l                 listen mode\n"
                 " -f                 accepts multible connects in listening mode\n"
                 " -z                 zero IO mode (used for scanning)\n"
        );
}

int main_loop (int local_in, int local_out, int sock) {
 int    len = 1;
 char   buf[1024];
 fd_set sl;
 struct timeval tv;
 int maxfh = (local_in > sock ? local_in : sock) + 1;

 while (len > 0) {
  FD_ZERO(&sl);
  FD_SET(local_in, &sl);
  FD_SET(sock, &sl);

  tv.tv_sec  = 1;
  tv.tv_usec = 0;

  if (select(maxfh, &sl, NULL, NULL, &tv) > 0) {
   if ( FD_ISSET(sock, &sl) ) {
    if ( (len = read(sock, buf, 1024)) == -1 )
     return -1;
    if ( write(local_out, buf, len) != len )
     return -1;
   } else {
    if ( (len = read(local_in, buf, 1024)) == -1 )
     return -1;
    if ( write(sock, buf, len) != len )
     return -1;
   }
  }
 }
 return 0;
}

char * localnode (void) {
 static char node[16] = {0};
 struct dn_naddr      *binaddr;
 struct nodeent       *dp;

 if ( !node[0] ) {
  if ( (binaddr=getnodeadd()) == NULL)
   return NULL;

  if ( (dp=getnodebyaddr((char*)binaddr->a_addr, binaddr->a_len, PF_DECnet)) == NULL )
   return NULL;

  strncpy(node, dp->n_name, 15);
  node[15] = 0;
 }

 return node;
}

int main (int argc, char * argv[]) {
 int sock;
 int i;
 int verbose   = 0;
 int zeroio    = 0;
 int listening = 0;
 int forking   = 0;
 char * k      = NULL;
 char * node   = NULL;
 char * object = NULL;
 int  objnum   = 0;
 struct sockaddr_dn sockaddr;
 socklen_t          socklen = sizeof(sockaddr);
 struct nodeent   * ne;

 progname = argv[0];

 for (i = 1; i < argc; i++) {
  k = argv[i];
  if ( strcmp(k, "-h") == 0 || strcmp(k, "--help") == 0 ) {
   usage();
   return 0;
  } else if ( strcmp(k, "-z") == 0 ) {
   zeroio = 1;
  } else if ( strcmp(k, "-l") == 0 ) {
   listening = 1;
  } else if ( strcmp(k, "-f") == 0 ) {
   forking = 1;
  } else if ( strcmp(k, "-v") == 0 ) {
   verbose++;
  } else if ( node == NULL ) {
   node = k;
  } else if ( object == NULL ) {
   object = k;
  } else {
   usage();
   return 1;
  }
 }

 if ( node == NULL ) {
  fprintf(stderr, "Error: need both, node and object name/number\n");
  usage();
  return 1;
 }

 if ( (k = strstr(node, "::")) != NULL ) {
  *k = 0;
  if ( k[2] != 0 && object == NULL )
   object = k+2;
 }

 if ( listening && node != NULL && object == NULL ) {
  object = node;
  node   = localnode();
 }

 if ( node == NULL || object == NULL ) {
  fprintf(stderr, "Error: need both, node and object name/number\n");
  usage();
  return 1;
 }

 if (*object == '#') {
  objnum = atoi(object+1);
 }

 if ( verbose ) {
  if ( listening ) {
   // listening on [any] 1234 ...
   fprintf(stderr, "listening on [any] %i (%s) ...\n", objnum, object);
  }
 }

 errno = 0;
 if ( listening ) {
  sock = dnet_daemon(objnum, object, 0, 0);
 } else {
  sock = dnet_conn(node, object, SOCK_STREAM, 0, 0, 0, 0);
 }

 if ( sock == -1 ) {
  // localhost [127.0.0.1] 23 (telnet) : Connection refused
  fprintf(stderr, "%s %i (%s): %s\n", node, objnum, object, strerror(errno));
  return 2;
 }

 if ( verbose ) {
  if ( ! listening ) {
   // localhost [127.0.0.1] 22 (ssh) open
   fprintf(stderr, "%s %i (%s) open\n", node, objnum, object);
  }
 }

 if ( listening ) {
  dnet_accept(sock, 0, NULL, 0);
  if ( verbose ) {
   memset(&sockaddr, 0, socklen);

   if ( getpeername(sock, (struct sockaddr*)&sockaddr, &socklen) == 0 ) {
    if ( sockaddr.sdn_objnum )
     sprintf((char*)sockaddr.sdn_objname, "#%u", sockaddr.sdn_objnum);

    ne = getnodebyaddr((char *)sockaddr.sdn_add.a_addr, sockaddr.sdn_add.a_len, PF_DECnet);
    fprintf(stderr, "connect to %s::%s from %s::%s \n", node, object, (char*)(ne == NULL ? NULL : ne->n_name), sockaddr.sdn_objname);
   } else {
    // connect to [127.0.0.1] from localhost [127.0.0.1] 53255
    fprintf(stderr, "connect to %s::%s from unknown \n", node, object);
   }
  }
 }

 if ( !zeroio )
  main_loop(0, 1, sock);

 if ( listening && !forking )
  kill(getppid(), SIGTERM);

 close(sock);

 return 0;
}

//ll
